README
Ada notes:
=========

This README file describes briefly the main directories of the Ada contrib.
The software compiles with Ada05 (Gnat GPL 2010) and runs on Unix (Linux).
Some programs depend on PCRE (Perl Compatible Regular Expressions) and work
with version 7.8 or above (8.11 is OK).
Programs are rated from 1 (simple) to 3 (very complex), on subjective criteria.

0. Gnatmake (optional)
-----------
The script ada/gnatmake can be set in the PATH (you can copy it or make a link
from $HOME/bin/gnatmake to it).
Similarly, the commands "ada" and "gnatlink" can be in the path and be this
gnatmake (you can copy it or make links from $HOME/bin/ada and
$HOME/bin/gnatlink to it).
Automatic generation by make (see next section) don't use them but these scripts
are convenient.

1. Makefiles project (MANDATORY)
--------------------
This project gathers several generic definitions for C and Ada generation.
Several examples of usage can be found in the source directories.

1.1 Relocation of the Makefiles directory
If the Makefiles installation directory is not $(HOME)/Makefiles,
- the definition of TEMPLATES in common.mk has to be adapted,
- all makefiles must have their first include directive of common.mk adapted.

1.2 Relocation of C or Repository
If the c or the repository directory are not in $(HOME)/ada, then the template
path.mk of Makefiles must be modified.

1.3 Recursive make
The content of the makefile for applying make to subdirecotries is:

SUBDIRS := c reposit usr
include $(HOME)/Makefiles/dir.mk

Other directives can be added if something has also to be compiled in current
dir.

2. C directory
--------------
Four libraries here:

2.1. Lib C util (cutil)
This lib gathers several low level utility libraries:
- timeval, used by other C libraries, for struct timeval operations,
- socket, interfaced by Ada Socket package, interfaces socket (tcp, udp, ipm,
  message on tcp, afinet or afux) calls,
- sys_calls, interfaced by Ada Sys_Calls and some other Ada packages, for
  several operating system calls,
- wait_evt, interfaced by Ada Even_Mng package, waits (with "select") on several
  fds, catches signals...

2.2 Lib X Mng (xmng)
This lib encapsulates all needed calls to X11, interfaced by X_Mng Ada package.

2.3 Lib C Ndbm (cnbdbm)
This lib interfaces with ndbm C library. The corresponding Ndbm package is not
used at present (except in test).

2.4 Lib Posix to PCRE (posic2pcre)
Own binding of PCRE to POSIX API, with extension, interfaced by
Regular_Expressions Ada package.

3. Reposit directory
--------------------
Many utility packages here. Some of them interdependant, the main dependency
trees are:
Graphic (X11) function: X_Mng <- (Genric_)Con_Io <- Afpx and Curve
Communication, Timers: Socket, Timers and Event_Mng <- Tcp_Util <- Channels 
See reposit/REPOSIT_LIST for the list of packages.

4. Usr directories
------------------

Account (3)
-------
Based on Afpx, this tool makes the management of bank accounts, with cheque,
credit card (defered) and transfer operations.

Agite (2)
-----
GUI for Git, dir oriented.

Als (3)
---
Lists, more or less the "ls" way, files and directories. Allow selection by
date and type, merge of all outputs, and sorting.

Approx (3)
------
Using Afpx and Curve, the program computes and displays the polynomial
approximation of a set of points in the plan.

Astub (3)
-----
Ada05 stubber (makes compilable empty body from spec).

Asubst (3)
------
Substitutes strings matching regexp by strings in files.

Chess (3)
-----
With Con_Io and Tcp_Util, allows 2 players to play chess on one or 2 computers.
Checks movement validity (already tricky) but does not play against human.

C_Magic (1)
-------
Small program that computes magic squares (numbers from to 1 in a square, with
same sum on all rows, cols and diags). Consumes a lot of CPU despite many
optims :-)

Code (1)
----
Simple program that codes (crypts) and decodes a text. The crypting algo is
very old and not secure.

Day_Of_Week (1)
-----------
Simple program that tells everything on today (or on another day).

Dictio (3)
------
Distributed data dictionnary that allows to store data, distribute this data
base on all the machines. Ensures correct insertion of new machines and
constant consistency. Recovery of data loss may be optimized (resends all).

Enigma (2)
------
Encodes/decodes text like german enigma machine did in ww2.

Fifos (2)
-----
Use Dictio to make distributed point to point connections (logical adress,
automatic connection and re-connection).

Fl (1)
--
Simple computation of fligh log (adding hours and minutes).

Great_Circle (1)
------------
Computation of heading and distance between two points on earth (GPS).

Heart (3)
-----
DB of heart rate measures, displayed with Afpx.

Hungar (2)
------
Euristic search (hungarian method) of best affection (e.g. of peaple to jobs).
Tough algo.

Lem (2)
---
Game: land the LEM on the moon.

Lsadeps (2)
-------
Shows the list/tree of Ada units/files dependencies

Mcd (3)
---
Reverse polish computer. At command line level but very powerful.

Misc
----
Many simple programs here. The most interesting are:
Alook (2) properly formats the words of a (valid) Ada05 source file.
Dtd_Checker (1) check a DTD file (of a xml).
Prime (1) search prime numbers.
Stat (1) measures the number of instruction of a (valid) Ada source file.
Status (1) evaluates if a target file needs to be rebuilt.
Tcping (1) pings in tcp a host/port.
Trail_spaces (1) removes tabs and trailing spaces and dos2unix a file.
Udp_Spy (2) displays udp/ipm packets received on a port.
Xml_Checker (2) uses Xml parser and generator (from reposit) to check XML.

Mmind (2)
-----
Mastermind game.

Navig (2)
-----
Navigation computation (wind, speed, heading, route, drift...).

Nimmari (2)
-------
Nim and Marienbad games.

Pexec (1)
-----
Recursive execution on directories of commands.

Pipe (1)
----
Uses Fifo to relay stdin to another host where another pipe can put it on
stdout.

Renardeau (2)
---------
Computes a target number by combining base numbers.

Simusi (2)
------
Machining simulation. (Tough algo).

Sokoban (3)
-------
The famous Sokoban game (push the boxes).

Tasks
-----
Several small programs playing with tasks.

Tests
-----
Several small test programs of the Repository packages.

Xwords (2)
------
A graphical lexicography of common words and nouns (french DB).

5. cots directory
-----------------
This directory mainly contains copies of:
- pragmarc and mine from http://pragmada.x10hosting.com/ (under GPL)
- LinXtris from http://sourceforge.net/projects/linxtris (under GPL)



HISTORY
Major changes:
2011/04/14: Add option --dotall (remove --display) that allows '.' to match '\n' in asubst, V12.0
            Add option Dot_All to Regular_Expressions
2011/04/04: Add support of port in http package and in tcpchat test scenario
2011/03/31: Add specific option 'E for entities expansion in xml_checker, V15.0
            Add normalize text also when knowing from DTD that there cannot be text in Xml_Parser, V26.6
2011/03/27: Fix homogenenize output of status in account
2011/03/24: Fix normalization of text and add adjustment of Node_Info in Xml_Parser, V26.5
2011/03/18: Fix preserve text when it is the only child in Xml_Parser, V26.2
2011/03/18: Add option --no-normalize to xml_checker, V14.0
2011/03/17: Chg default 'keep' mode of xml_checker, V13.0
2011/03/15: Fix autobus local echo and clean-up
            Chg replace Fifo by an Autobus in mcd
2011/03/11: Fix blocking management in socket and Tcp_Util.Abort_Send_And_Close
2011/03/10: Add operations to set/get TTL on socket
2011/03/09: Fix detection of '^' and '$' in search pattern when specific delimiter in asubst, V11.4
2011/03/08: Fix handling of specific Delim in asubst, V11.3
2011/03/01: Add different blocking modes for sending and receiving on sockets
2011/02/28: Add timeout in Tcp_Util.Send
2011/02/22: Add autobus package
2011/02/16: Add improve help of als, V6.2
2011/02/15: Add option --date_iso for strict ISO output of dates in als, V6.1
            Chg use ISO 8601 format for time input and output of als, V6.0
            Add a static agite executable
2011/02/14: Fix dispatching of X Refresh event to multiple windows
            Add Get_Clock in Chronos, and remove Detach cause same as Attach (null)
2011/02/13: Add use anonymous access type in Limited_List
2011/02/12: Add make Timer_Id a smart reference
2011/02/11: Add new operations on Timers and add New_Id parameter to expiration callback
2011/02/10: Add improve error reported when parsing encoding map in Xml_Parser V26.1
2011/02/08: Add operation Set_Pi to Xml_Parser.Generator V9.0
2011/02/07: Add make Timer_Id tagged
            Chg remove Text_Handler
2011/02/06: Add fix size of X window in window manager
2011/02/04: Add make State_Machine OO
2011/02/02: Add support for http://<ip_address>
2011/02/01: Fix in http the decoding when server replies an error (no header)
2011/01/31: Add make Ada_Parser and Computer re-entrant
2011/01/28: Add Hosting in github
2011/01/21: Add Set_Null in As.U and Unbounded_Arrays and use them
2011/01/20: Fix many small fixes in select_file and initializations of chess, approx and day_of_week
2011/01/07: Fix remove "use" of As.U and As.B
            Fix use As.B instead of Text_Handler and restore original Text_Handler
2010/12/28: Fix unbounded Tail when no pad needed
2010/12/22: Add our own (tagged) version of unbounded and bounded strings
2010/12/19: Add Move to first/last significant char on Ctrl-Left/Right in Afpx.Put_Then_Get
2010/12/16: Fix remove fixed-length strings from Reposit and make Text_Handler.Text tagged
2010/11/19: Add default background color in Con_Io and Afpx and change default to the "grey" scheme
            Fix font loading
2010/11/14: Add support of 4 digits in renardeau computation and result
2010/11/12: Fix dummy signal didn't wake up the wait
2010/11/11: Add replace Unique_List by Hashed_List.Unique, more efficient
2010/11/08: Add search of anagrams in xwords
2010/11/06: Add Hashing on larger keys and optimizations
2010/10/31: Add X screen to renardeau, move it in a separate dir
2010/10/30: Add program renardeau in misc
2010/10/27: Add error handler of external commands of tcpchat
2010/10/21: Add IfUnset trilean to if/elsif/while of tcpchat
2010/10/19: Add conditional setting and computation of variables in tcpchat
2010/10/18: Add new operation Repeat to tcpchat
2010/10/17: Add new operations (Set, Cond...) to tcpchat
2010/10/01: Add program tcpchat that plays chat scripts on a TCP connection
2010/08/29: Add Sys_Calls operation to set/reset Close_On_Exec on File_Desc
2010/08/23: Add support of URIs "file://" and "http://" in Xml_Parser V24.0
2010/08/22: Fix socket C read that shall return read length (i.o. SOC_OK) on tcp_socket (no header)
2010/08/20: Add Change interfaces of Con_Io and Afpx to use Unicode instead of Wide_Char
2010/08/18: Add Tcp_Util end of overflow callback shall return True to Event_Mng
2010/08/17: Add support of UTF-8 input in Async_Stdin
2010/08/13: Add support of tcp channel in mcd
2010/08/12: Add support of udp/ipm channel in mcd (tcp support is on the way)
            Add \P...\p in replace string for file insertion of asubst V11.0
            Add Remove string len limitation of Argument
            Add Make Socket_Dscr tagged and use object call notation
            Add Improve definition of colors in Afpx.xml
            Add a button "revert" and a history "scroll bar" in agite V2.8
2010/07/28: Add replace Queues.Timed.Get by Read
            Fix init of State_Machine to first state
            Add Home1 and End1 sequences in Async_Stdin
            Add support of mode Inout_File in Text_Line
            Add restore of current entry on Reread in Select_File
            Add declare Real_Io in My_Math
            Fix Delete_Connection in Network 
2010/07/23: Fix C_Nbres.Polar constructor when negative imaginary part
            Fix decoding of Utf_16 when one char
            Fix Append of Unbounded_Arrays
2010/07/17: Fix update of current dir field in agite V2.4
            Fix use path of XML for DTD during Check in Xml_Parser V23.4
            Add complete handling of bookmarks in agite V2.3
            Fix handling in del_fd when last_fd>32
            Fix close of dtd and external entity files in Xml_Parser V23.3
2010/07/15: Add Dur_Image
2010/07/12: Fix handling of empty dtd file name in Xml_Parser V23.2
            Fix several reference to Xxx_Gray color
2010/07/11: Add replace Ada.Strings.Unbounded by As.U everywhere

c/socket.h
#ifndef __SOCKET_H__
#define __SOCKET_H__

#include <stdio.h>
#include "boolean.h"

/* Variables and types */
/* ------------------- */

/* OK */
#define SOC_OK        0

/* Errors, see also soc_error() */
/* Socket not open, already open */
#define SOC_USE_ERR     -1
/* System error (traced and errno set) */
#define SOC_SYS_ERR     -2
/* Dest set or not set */
#define SOC_DEST_ERR    -3
/* Linked or not linked */
#define SOC_LINK_ERR    -4
/* Connected or not connected */
#define SOC_CONN_ERR    -5
/* Bcast when not allowed by proto */
#define SOC_BCAST_ERR   -6
/* Length to short */
#define SOC_LEN_ERR     -7
/* Set_for_reply set or not set */
#define SOC_REPLY_ERR   -8
/* Called soc_send instead of soc_resend despite prev send returned SOC_WOULD_BLOCK */
#define SOC_TAIL_ERR    -9
/* Call not allowed for this protocol */
#define SOC_PROTO_ERR  -10
/* Close while fd used in select */
#define SOC_FD_IN_USE  -11
/* Lan/host address or port string format error */
# define SOC_FMT_ERR   -12

/* Failures */
/* Connection refused */
#define SOC_CONN_REFUSED   -21
/* Get*by*: entry not found */
#define SOC_NAME_NOT_FOUND -22
/* Operation would block - do it again */
#define SOC_WOULD_BLOCK    -23
/* Connection lost */
#define SOC_CONN_LOST      -24
/* Address in use. Close-wait? */
#define SOC_ADDR_IN_USE    -25
/* Read -> 0 : disconnection if after a select */
#define SOC_READ_0         -26
/* Set_for_reply could not set the ipm sending interface */
#define SOC_REPLY_IFACE    -27

/* String terminator */
#define NUL '\0'

/* Byte of 8 bits.  Values from 0 to   255. */
#define byte unsigned char

/* Word of 16 bits. Values from 0 to 65535. */
#define word unsigned short int

/* Protocols */
typedef enum {udp_socket=0, tcp_socket=1, tcp_header_socket=2,
              tcp_afux_socket=3, tcp_header_afux_socket=4} socket_protocol;

/* Blocking mode */
/* full_blocking: soc_accept, soc_receive, soc_send, soc_resend and */
/*                soc_set_dest_xxx are blocking */
/* blocking_send: soc_accept and soc_receive are non blocking */
/* non_blocking: connecting, sending, accepting and receiving are non */
/*               blocking */
/* The mode where only reception is blocking is not usefull */
typedef enum {full_blocking, blocking_send, non_blocking} blocking_mode;

/* Token socket : a pointer (abstract data type) */
/* Has to be initialised to init_soc */
typedef void * soc_token;
#define init_soc  ((soc_token*) NULL)

/* Port of socket : a word */
#define soc_port word

/* Host address (network format) */
typedef union soc_host_t {
    /* 4 bytes in network order */
    unsigned int integer;
    byte bytes[4];
} soc_host;

/* String length for a host or a port */
#define SOC_MAX_LEN 50

/* A message */
typedef int soc_length;
typedef void * soc_message;

/* Returns the string associated to an error code */
extern const char * soc_error (const int code);

/*----------------------------------------------------------------*/
/* All functions return an error or SOC_OK except soc_receive:    */
/*  on non full_blocking tcp no header it returns the read length */
/*  on udp it returns the real length of the message sent         */
/*----------------------------------------------------------------*/

/*-------------------------------------------------------------------------*/
/*  Note for Multicast IP (using Udp socket):                              */
/*  For sending IPM, simply Set_Destination to a LAN name                  */
/*    which is defined with a D class address, and a port.                 */
/*  For receiving IPM, first Set_Destination to the LAN name and port,     */
/*    then Link to the same port as this destination.                      */
/*    It is possible to link to a dynamic port (then port set in           */
/*    destination port is    not used).                                    */
/*  By default IPM emission and reception are from/to the "natural"        */
/*    interface (local host id). A specific interface can be specified     */
/*    before setting destination / before linking to port. If interface is */
/*    set for reception, then the set_for_reply flag of soc_receive sets   */
/*    the sending interface to it.                                         */
/*-------------------------------------------------------------------------*/


/* ------------------*/
/* General functions */
/* ------------------*/
/* Open a socket (in blocking mode) */
extern int soc_open (soc_token *p_token, socket_protocol protocol);

/* Close a socket */
extern int soc_close (soc_token *p_token);

/* Gets the id of a socket (for select, ioctl ... ) */
extern int soc_get_id (soc_token token, int *p_id);

/* Set the socket blocking or non blocking */
/*  (for connecting and sending, for receiving) */
/* Socket is full_blocking at creation (open/accept) */
extern int soc_set_blocking (soc_token token, blocking_mode blocking);

/* What is the socket blocking mode */
extern int soc_get_blocking (soc_token token, blocking_mode *blocking);

/* Get socket protocol */
extern int soc_get_protocol (soc_token token, socket_protocol *protocol);

/* Set the TTL. Socket must be either udp with dest set (otherwise */
/*  SOC_DEST_ERR), or tcp non afux (otherwise SOC_PROTO_ERR) */
extern int soc_set_ttl (soc_token token, byte ttl);

/* Get the TTL. Socket must be either udp or tcp non afux (otherwise */
/*  SOC_PROTO_ERR) */
extern int soc_get_ttl (soc_token token, byte *ttl);

/*-------------------------------------*/
/* Emission                            */
/* No broadcast nor change dest in tcp */
/* Socket must not be linked in tcp    */
/*-------------------------------------*/

/* Set the interface on which send mutlicast IP (udp_socket). */
/* Set it before setting destination. */
/* 0 to reset */
/* Beware that setting the sending interface is not always supported */
/*  and may require to be root (Set_Destination will return SOC_SYS_ERR). */
extern int soc_set_send_ipm_interface (soc_token token, const soc_host *host);

/* Set the destination host/lan name and port - specify service */
/* Broadcast if lan */
extern int soc_set_dest_name_service (soc_token token, const char *host_lan,
                                      boolean lan, const char *service);

/* Set the destination host name and port - specify port */
/* Broadcast if lan */
extern int soc_set_dest_name_port (soc_token token, const char *host_lan,
                                   boolean lan, soc_port port);

/* Set dest to a host, service */
extern int soc_set_dest_host_service (soc_token token, const soc_host *host,
                                      const char *service);
/* Set dest to a host, port */
extern int soc_set_dest_host_port (soc_token token, const soc_host *host,
                                   soc_port port);

/* Change destination host_lan name (same port) */
/* Destination must have been previously set (by a set or a rece) */
/* Broadcast if lan */
extern int soc_change_dest_name (soc_token token, const char *host_lan,
                                 boolean lan);

/* Change destination host (same port) */
extern int soc_change_dest_host (soc_token token, const soc_host *host);

/* Change destination port (same host_lan) - specify service */
extern int soc_change_dest_service (soc_token token, const char *service);

/* Change destination port (same host_lan) - specify port */
extern int soc_change_dest_port (soc_token token, soc_port port);

/* Get the destination port */
extern int soc_get_dest_port (soc_token token, soc_port *p_port);

/* Get the destination host */
extern int soc_get_dest_host (soc_token token, soc_host *p_host);

/* Send to a socket, the destination of which must be set */
/* May return SOC_WOULD_BLOCK, then next tries have to be made */
/*  with soc_resend util it returns ok */
extern int soc_send (soc_token token, soc_message message, soc_length length);

/* Resend tail of previous message */
extern int soc_resend (soc_token token);

/*---------------------------------------------*/
/* Search in local host/port/network databases */
/*---------------------------------------------*/
/* Get current lan name (computed from local host name) */
/* lan_name must be large enough */
extern int soc_get_lan_name (char *lan_name, unsigned int lan_name_len);

/* Find name of soc_host and vice versa */
extern int soc_host_name_of (const soc_host *p_host, char *host_name,
                             unsigned int host_name_len);
extern int soc_host_of (const char *host_name, soc_host *p_host);

/* Find name of soc_port and vice versa */
extern int soc_port_name_of (const soc_port port,
                             const socket_protocol proto,
                             char *port_name,
                             unsigned int port_name_len);
extern int soc_port_of (const char *port_name,
                        const socket_protocol proto,
                        soc_port *p_port);

/* Gets local host */
extern int soc_get_local_host_name (char *host_name,
                                    unsigned int host_name_len);
extern int soc_get_local_host_id (soc_host *p_host);

/* String "x.y.z.t" to host, and string to port conversions */
extern int soc_str2host (const char *str, soc_host *p_host);

extern int soc_str2port (const char *str, soc_port *p_port);

/* ------------------------------------*/
/* Receive                             */
/* Socket must not be connected in tcp */
/* ------------------------------------*/

/* Set the interface on which receive mutlicast IP (udp_socket). */
/* Set it before linking. */
/* 0 to reset */
extern int soc_set_rece_ipm_interface (soc_token token, const soc_host *host);

/* Links the socket to a port specified by the service */
/* The socket must be open and not already linked */
extern int soc_link_service (soc_token token, const char *service);

/* Links the socket to a port specified by it's value */
/* The socket must be open and not already linked */
extern int soc_link_port (soc_token token, soc_port port);

/* Links the socket to a port dynamically choosen by the system */
/* The socket must be open and not already linked */
extern int soc_link_dynamic (soc_token token);

/* Gets the port to which is linked a socket */
extern int soc_get_linked_port (soc_token token, soc_port *p_port);

/* To test if there is a message to receive and receive it */
/* Length must be initialized with the size of the buffer */
/* The socket must be open, linked in udp and not linked in tcp */
/*  After success, the socket may be ready for a send to reply */
/* No set_for_reply if tcp */
/* In IPM, the interface for sending may also be set but see */
/*  Set_Sending_Ipm_Interface. */
/*  If set_for_reply is FALSE then set_ipm_iface is not significant. */
/* Returned values: */
/*   - the length of bytes read, which is usually the length of the message */
/*     sent. In tcp no header it may be anything from 0 to length. */
/*     In udp it is the sent message length (maybe > length, or 0). */
/*   - SOC_WOULD_BLOCK (in non blocking), new read has to be done */
/*   - SOC_READ_0, tcp disconnection? */
/*   - any other (fatal) error */
extern int soc_receive (soc_token token,
                        soc_message message, soc_length length,
                        boolean set_for_reply,
                        boolean set_ipm_iface);

/*--------------------*/
/* Tcp specific calls */
/*--------------------*/

/* Accept a connection. */
/* The socket must be open, tcp or tcp_header and linked */
/* A new socket is created (tcp or tcp_header, same as token) with dest set */
/* May return SOC_WOULD_BLOCK (either in blocking OR non blocking) if no    */
/*  valid connection to accept. Go back to wait. */
extern int soc_accept (soc_token token, soc_token *p_token);

/* Is a tcp socket connected following a soc_set_... or a soc_accept */
extern int soc_is_connected (soc_token token, boolean *p_connected);


/* __SOCKET_H__ */
#endif

